sql: add support for ON UPDATE CASCADE for foreign key references

BramGruneir opened the pull request:
Enables the addition of the ON UPDATE CASCADE action for foreign keys references.

Major changes in `rowwriter.go`:
* Added a quick check to see if a cascader is required at all (this code is in `cascader.go`). And if not, the normal path is followed.
* When a cascader is required, UpdateRow can only use a single batch per row. Unless we can read from batches without running them, this will be a requirement going forward.
* Added the ability to skip foreign key checks for `InsertRow`.
* Merge the `DeleteRow()` And `deleteRowWithoutCascade()` functions.

Major changes in `cascader.go`:
* Added the new `makeCascader` functions that check to see if a cascader is required at all.
* Of course, added `updateRows()`, which performs all the updates.
* Added the checking of foreign key constraints at the end of a `cascadeAll()` call. This of course

Unlike with deletes, there is more to keep track of when looking for orphaned rows. In the wost case, rows can be updated multiple times. So row A -> B, then B -> C, then C -> D. But we don't want to test the middle states for foreign key violations, and we only want to test A -> D. So this accomplished by storing all transitions and looking forward through these updates to find if a row was updated again. There is potential to improve this using a map of some sort, but that can be done in a further update. The normal case is that there is only a single update and there is a quick path to only check that.

There is also a need to compare the contents two rows. Using the example above, say after A -> B and then B -> C, there needs to be a way to determine that the first B is the equivalent of the second B. To accomplish this, a new function on `tree.Datums` `IsDistinctFrom()` was added that treats nulls as equivalent.

Release note (SQL): ON UPDATE CASCADE foreign key constraints are fully supported

BramGruneir committed:
sql: add support for ON UPDATE CASCADE for foreign key references

Enables the addition of the ON UPDATE CASCADE action for foreign keys references.

Major changes in `rowwriter.go`:
* Added a quick check to see if a cascader is required at all (this code is in `cascader.go`). And if not, the normal path is followed.
* When a cascader is required, UpdateRow can only use a single batch per row. Unless we can read from batches without running them, this will be a requirement going forward.
* Added the ability to skip foreign key checks for `InsertRow`.
* Merge the `DeleteRow()` And `deleteRowWithoutCascade()` functions.
* Pass the evalContext down into the row writer.

Major changes in `cascader.go`:
* Added the new `makeCascader` functions that check to see if a cascader is required at all.
* Of course, added `updateRows()`, which performs all the updates.
* Added the checking of foreign key constraints at the end of a `cascadeAll()` call.

This PR follows a similar route to the one that added ON DELETE CASCADE. The row updater creates a cascader and after performing its initial update operation, calls cascadeAll() to perform the cascading updates.  Inside the cascader, the index selection and cascading queue were already setup, so it was relatively easy to setup updateRows and get it to act in a similar way. However, this is where problems started to creep up immediately.

Unlike the deleter, if the original changes have not been run (as in the batch has been run), then future updates cascading values will not be found.  Since batches cannot be read from, the initial changes must be run prior to calling into the cascader.  This is problematic as it will adversely affect the speed of any non-cascading updates since each individual update will have to be run in a single batch. To get around this, the makeUpdateCascader function was added that only creates a cascader if a cascading update is possible. One was added for deleting as well. If there is no cascader the original batch is used and the cascader is never called into.

When cascading a delete, each delete has two steps. First, fetch the primary key columns of the rows that need to be deleted. Second, fetch those rows, specifically only the columns required for the rowDeleter and then delete them. With updates, I mistakenly made the assumption that it would be possible to eschew the first lookup and to just cascade the update directly. This proved to be incorrect. Not only are the primary key columns required, but all columns that have foreign key constraints are.

Unlike with deletes, there is more to keep track of when looking for orphaned rows. In the worst case, rows can be updated multiple times. So let say we have row A, and it gets updated 3 times.  Row A -> B, then B -> C, and then C -> D. But we don't want to test the middle states for foreign key violations as they will always fail. We only want to test A -> D. It's important to note, that relying on primary keys is not good enough, as even they can be updated.

Sadly, there is no easy way of accomplishing this. After trying numerous iterations on how to perform this check, I settled on just storing all transitions as a tuple based on two row containers per table. Each transition has an entry in both, one for the original and one for the updated value. Then by performing a forward look through all transitions pairs to find if a row was updated again. The usual case is that there is only a single update and there is a quick path to only check that. This is where the requirement to compare the full rows was needed. Unless all values that could be updated are stored, it would not be not possible to compare two versions of the same row.

But this leads to yet another issue.  Now there was a need to be able to compare the contents of two rows. Using the example above, say after A -> B and then B -> C, there needs to be a way to determine that the first B, is the equivalent of the second B. To accomplish this, a new function on `tree.Datums` `IsDistinctFrom()` was added. This then, based on review comments, leads to the need to pass an evalContext all the way down into the cascader.  This was relatively simple but required a little bit of plumbing.

Release note (SQL): ON UPDATE CASCADE foreign key constraints are fully supported

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21329)
<!-- Reviewable:end -->

knz commented:
Very nice work. I like how well polished it is.

As a reviewer it is still hard to review this successfully however, due to the complex nature of the change. I would have preferred if the commit message could walk me through the story of what is being changed and why. Right now the commit message reads more like a description of the patch "after the fact". I'd rather have a story tell me "so here's the situation before this patch. To achieve on update cascade, we need to ensure X and Y. The place to do this is Z. So the patch does A and B in location Z, but then we also need to do C and D in location W." And so forth.

Also there's a typo in the commit message: wost -> worst.

---

Reviewed 14 of 14 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 95 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W2Z-VvPxY8Vp953U5:-L2W2Z-VvPxY8Vp953U6:b-svoldg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/logictest/testdata/logic_test/cascade#L95)):*
> ```
> 
> query I
> SELECT COUNT(*) FROM a
> ```

compress this: 
```
query IIIIII
select 
    (select count(*) from a),
    (select count(*) from b1),
    (select count(*) from b2),
    (select count(*) from c1),
    (select count(*) from c2),
    (select count(*) from c3)
----
0 0 0 0 0
```

here and below

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 1150 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W33xttekvIm0GKNde:-L2W33xttekvIm0GKNdf:blvjmu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/logictest/testdata/logic_test/cascade#L1150)):*
> ```
> 
> statement ok
> INSERT INTO b1 VALUES ('b1-pk1', 'original');
> ```

Make a single block, also use a single insert for adding multiple rows to the same table:

```
statement ok
INSERT INTO b1 ...; 
   INSERT INTO b2 ...;
   INSERT INTO c1 VALUES (...), (...), ...;
   INSERT INTO c2 VALUES (...), (...), ...
```

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 1643 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W3L0VAYiUMNMYiQ6s:-L2W3L0VAYiUMNMYiQ6t:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/logictest/testdata/logic_test/cascade#L1643)):*
> ```
> 
> statement ok
> INSERT INTO a VALUES ('original'), ('updated');
> ```

ditto

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 1706 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W3LzIDEoTHKPMwVuQ:-L2W3LzIDEoTHKPMwVuR:b95bt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/logictest/testdata/logic_test/cascade#L1706)):*
> ```
> 
> # Clean up after the test.
> statement ok
> ```

compress: `drop table c3, c2, c1, b2, b1, a`

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 2051 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W3dis21mWgse0bK3b:-L2W3dis21mWgse0bK3c:b-irkbl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/logictest/testdata/logic_test/cascade#L2051)):*
> ```
> 
> query T
> SELECT * FROM a
> ```

compress: `select * from a, b, c, d, e, f` -- with 1 row in each table you should get just 1 row in the cross product

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 2168 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W3k8rYub9kftXy4Ko:-L2W3k8rYub9kftXy4Kp:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/logictest/testdata/logic_test/cascade#L2168)):*
> ```
> 
> query T
> SELECT * FROM a
> ```

ditto

---

*[pkg/sql/sem/tree/datum.go, line 179 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W3r-vYoJBhZSUmz7w:-L2W3r-vYoJBhZSUmz7x:b-5fkhej) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/sem/tree/datum.go#L179)):*
> ```Go
> 			}
> 		} else {
> 			if val.Compare(evalCtx, other[i]) != 0 {
> ```

I fear there are some data types whose Compare method needs a non-default EvalContext (ts with timestamp, collated strings). Any way to pass the evalctx as argument to this function? And arrange it to be proper in the caller?

---

*[pkg/sql/sqlbase/cascader.go, line 612 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W4iWEW3rsPfOBuz2p:-L2W4iWEW3rsPfOBuz2q:b-luxxof) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/sqlbase/cascader.go#L612)):*
> ```Go
> 	if traceKV {
> 		log.VEventf(ctx, 2,
> 			"cascading update from refIndex:%s, into table:%s, using index:%s from values:%s to values:%s",
> ```

If you mean this to appear in the output of `show kv trace` (I suppose you do, otherwise you wouldn't make it conditional on `traceKV`), you need to add the corresponding filtering condition in `show_trace.go`.

---

*[pkg/sql/sqlbase/cascader.go, line 892 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W55okUPd0MQqYGjLw:-L2W55okUPd0MQqYGjLx:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/sqlbase/cascader.go#L892)):*
> ```Go
> 			if elem.updatedValues != nil {
> 				log.VEventf(
> 					ctx, 2, "cascading into %s for original values:%s updated to:%s",
> ```

ditto

---

*[pkg/sql/sqlbase/rowwriter.go, line 630 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W5DaaqPafGgsKR66U:-L2W5DaaqPafGgsKR66V:b-bs4if9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/sqlbase/rowwriter.go#L630)):*
> ```Go
> 	traceKV bool,
> ) ([]tree.Datum, error) {
> ```

remove the empty newline

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-:-L2W5OvdlIf9tAzpK8De:b-lppnfh)*
<!-- Sent from Reviewable.io -->

dt commented:
Reviewed 13 of 14 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/csv.go, line 606 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2MgPva5dnlQLMfvQ_h:-L2MgPva5dnlQLMfvQ_i:bz113wl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/ccl/sqlccl/csv.go#L606)):*
> ```Go
> 				row,
> 				true, /* ignoreConflicts */
> 				sqlbase.CheckFKs,
> ```

Hm. IIRC we reject FKs in schemas passed to IMPORT so I think this might not matter in practice, but the we definitely aren't actually checking existing rows while constructing these SSTs (like, I don't think they have a real txn to work with or anything).

---

*[pkg/ccl/sqlccl/load.go, line 305 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2MgQt3FqWe9do7eBPf:-L2MgQt3FqWe9do7eBPg:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/ccl/sqlccl/load.go#L305)):*
> ```Go
> 			return errors.Wrapf(err, "process insert %q", row)
> 		}
> 		if err := ri.InsertRow(ctx, b, row, true, sqlbase.CheckFKs, false /* traceKV */); err != nil {
> ```

ditto

---

*[pkg/sql/sem/tree/datum.go, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2MfFY2TtSoo7pM4D8_:-L2MfFY2TtSoo7pM4D8a:b2xn9k4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/sem/tree/datum.go#L172)):*
> ```Go
> 		return true
> 	}
> 	evalCtx := &EvalContext{}
> ```

Huh, I haven't been keeping careful tabs on sql eval lately, so I have no idea, but is it correct to just make an empty one of these?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Ok, all comments addressed and I've spent some time trying to describe what changes where made and why.  Let me know if it helps.  

I think I might be able to remove the passing of the bytes monitor down since it's part of the evalCtx now, but I'll do that in a follow up PR.

---

Review status: 2 of 18 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/ccl/sqlccl/csv.go, line 606 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2MgPva5dnlQLMfvQ_h:-L2WDLeKTM4bwKiQaiRv:b7tuvkd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/ccl/sqlccl/csv.go#L606)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Hm. IIRC we reject FKs in schemas passed to IMPORT so I think this might not matter in practice, but the we definitely aren't actually checking existing rows while constructing these SSTs (like, I don't think they have a real txn to work with or anything).
</blockquote></details>

In this case, since the only option was to check the FKs, I just left it as is.  I'll add a TODO to investigate if that's required here.

---

*[pkg/ccl/sqlccl/load.go, line 305 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2MgQt3FqWe9do7eBPf:-L2WDbxOpCQ-W2Oa0jfV:b-udxhft) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/ccl/sqlccl/load.go#L305)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

also added a todo here.

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 95 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W2Z-VvPxY8Vp953U5:-L2WMpWMQQ0xBAyDz4S3:bewwfku) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/logictest/testdata/logic_test/cascade#L95)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

compress this: 
```
query IIIIII
select 
    (select count(*) from a),
    (select count(*) from b1),
    (select count(*) from b2),
    (select count(*) from c1),
    (select count(*) from c2),
    (select count(*) from c3)
----
0 0 0 0 0
```

here and below

</blockquote></details>

That's a lot cleaner.  Done.

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 1150 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W33xttekvIm0GKNde:-L2WWvMN1iD6EsEkMc_J:b-psrf5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/logictest/testdata/logic_test/cascade#L1150)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Make a single block, also use a single insert for adding multiple rows to the same table:

```
statement ok
INSERT INTO b1 ...; 
   INSERT INTO b2 ...;
   INSERT INTO c1 VALUES (...), (...), ...;
   INSERT INTO c2 VALUES (...), (...), ...
```
</blockquote></details>

done.  This looks a lot better and shrunk this file down a lot. 

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 1643 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W3L0VAYiUMNMYiQ6s:-L2WY3sy0zhWiiMN3Ogy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/logictest/testdata/logic_test/cascade#L1643)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 1706 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W3LzIDEoTHKPMwVuQ:-L2WYEQJgrRuvqkSQQWN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/logictest/testdata/logic_test/cascade#L1706)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

compress: `drop table c3, c2, c1, b2, b1, a`
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 2051 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W3dis21mWgse0bK3b:-L2WZnvuYvTJO1RG9jvz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/logictest/testdata/logic_test/cascade#L2051)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

compress: `select * from a, b, c, d, e, f` -- with 1 row in each table you should get just 1 row in the cross product
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 2168 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W3k8rYub9kftXy4Ko:-L2WZpEc62mjWBAdvKFB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/logictest/testdata/logic_test/cascade#L2168)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/datum.go, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2MfFY2TtSoo7pM4D8_:-L2xoxrf2zPVjovhaA_A:b-hepms7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/sem/tree/datum.go#L172)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Huh, I haven't been keeping careful tabs on sql eval lately, so I have no idea, but is it correct to just make an empty one of these?
</blockquote></details>

After consultations with @andreimatei, you are correct and the eval context is now passed down into the cascader.

---

*[pkg/sql/sem/tree/datum.go, line 179 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W3r-vYoJBhZSUmz7w:-L2xpISDzLKS05ScVYmU:b-r6v3pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/sem/tree/datum.go#L179)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I fear there are some data types whose Compare method needs a non-default EvalContext (ts with timestamp, collated strings). Any way to pass the evalctx as argument to this function? And arrange it to be proper in the caller?
</blockquote></details>

Yep, done.  It was easier than I thought.

---

*[pkg/sql/sqlbase/cascader.go, line 612 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W4iWEW3rsPfOBuz2p:-L2Wjo6knfH_9TZ_eZ-L:bs6yxif) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/sqlbase/cascader.go#L612)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

If you mean this to appear in the output of `show kv trace` (I suppose you do, otherwise you wouldn't make it conditional on `traceKV`), you need to add the corresponding filtering condition in `show_trace.go`.
</blockquote></details>

Ah, excellent.  Done and it's a lot cleaner now and I added a test for both update and delete cascading.

---

*[pkg/sql/sqlbase/cascader.go, line 892 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W55okUPd0MQqYGjLw:-L2WsS2lqHTgBb-KqRqJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/sqlbase/cascader.go#L892)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/rowwriter.go, line 630 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2W5DaaqPafGgsKR66U:-L2W_MtFVOpG0kw_hu1X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9e0c6190996c7098c2a45c8f80cecbddbe8e863/pkg/sql/sqlbase/rowwriter.go#L630)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

remove the empty newline
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-:-L2xpQCO7BS_Lb659xbo:bytj7m7)*
<!-- Sent from Reviewable.io -->

knz commented:
Yes your commit message is now astonishingly good. 

Of course with such quality content it becomes possible to formulate follow-up concerns/questions. With the overhead of storing two copies of the tuples in memory, the cost of a multi-row update (when there are cascade clauses specified) becomes non-trivial.

I think that either in this PR or a follow-up you should create an additional benchmark in `pkg/bench` that exercises this code path and measures memory+time consumption for a simple combination of tables, with one benchmark doing the updates with ON CASCADE and one issuing multiple separate DML statements to do the same work in the same transaction.

knz commented:
(I forgot to mention, in case you also had forgotten you can get the mem stats with `TESTFLAGS=-benchmem`)

Then post the bench difference (as computed by `benchstat`) in the commit message for the commit where you also add the benchmark.

knz commented:
Oh and just a nit: can you please wrap your commit messages to 80 (or better, 72) columns.
Otherwise `git log` in a terminal will become obnoxious.

BramGruneir commented:
I'm planning on adding just such a benchmark.  But the business need of getting the cascading operations working trumped the benchmarking for now.

I've never had anyone else complain about line length in commits before.  I'll find a wrapper for all future ones.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-:-L2zgPAx8hY8zihu8Z_7:b-g3u07i)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
Just a nit about the trace message.

Well done!

---

Reviewed 16 of 16 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/show_trace.go, line 97 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2zokLM9JyEIdqFRp5K:-L2zokLM9JyEIdqFRp5L:b-xfpqj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/828438401286f582f60a2c0046f043b4ca611904/pkg/sql/show_trace.go#L97)):*
> ```Go
>    OR message LIKE 'execution failed: %'
>    OR message LIKE 'r%: sending batch %'
>    OR message LIKE 'Cascading %'
> ```

we don't usually do capitals in log/error messages. It makes less them composable with errors.Wrapf().

---

*[pkg/sql/sqlbase/cascader.go, line 615 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2zpAxMbZwFNOHUJo9F:-L2zpAxMbZwFNOHUJo9G:b5jz6ox) ([raw file](https://github.com/cockroachdb/cockroach/blob/828438401286f582f60a2c0046f043b4ca611904/pkg/sql/sqlbase/cascader.go#L615)):*
> ```Go
> 	// Create the span to search for index values.
> 	if traceKV {
> 		log.VEventf(ctx, 2, "Cascading update into table: %d using index: %d",
> ```

see my comment about capitals

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-:-L2zpdZBdJrQQtkpd1_f:b-hahkdq)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
TFTRs!  I'm excited to get this in.

---

Review status: 15 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/show_trace.go, line 97 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2zokLM9JyEIdqFRp5K:-L3-7WwY9J_cLRRbKRqj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/828438401286f582f60a2c0046f043b4ca611904/pkg/sql/show_trace.go#L97)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

we don't usually do capitals in log/error messages. It makes less them composable with errors.Wrapf().
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/cascader.go, line 615 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L2zpAxMbZwFNOHUJo9F:-L3-7Xbo2HbU0ZHrHJB7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/828438401286f582f60a2c0046f043b4ca611904/pkg/sql/sqlbase/cascader.go#L615)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

see my comment about capitals
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-:-L3-7YuHH7V0O7fmIKSa:bwmpk64)*
<!-- Sent from Reviewable.io -->

dt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 15 of 18 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-:-L3-9Sags65ywmexfa15:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 15 of 18 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/cascade, line 854 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329#-L51pf1Es8LjyS_qBfPV:-L51pf1Es8LjyS_qBfPW:b-ygfwo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/50b6aed77c0f324b31051eba651ec53cb2a631e1/pkg/sql/logictest/testdata/logic_test/cascade#L854)):*
> ```
> 
> # Update again but this time check show trace
> statement ok
> ```

Bram, this line is an accident, right?
If so, I'm grateful for it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21329)*
<!-- Sent from Reviewable.io -->
